### PR TITLE
feat: Support XZ decompression

### PR DESCRIPTION
**Pull request by yujincheng08**

### CODE REVIEW

File: src/read/xz.rs [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+impl<'a> Write for BufWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.inner.len() > *self.written {
+            let len = std::cmp::min(buf.len(), self.inner.len() - *self.written);
+            self.inner[*self.written..*self.written + len].copy_from_slice(&buf[..len]);
+            *self.written += len;
+            *self.total += buf.len();
+            Ok(len)
+        } else {
+            self.rest.extend(buf.iter().copied());
```
Comment (Pr0methean) [inline src/read/xz.rs:71]: How does `self.rest` ever get cleared?

File: src/read/xz.rs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+            count: &mut self.stream_size,
+        };
+        if *reader.count == 0 {
+            let mut b = [0u8; 6];
+            match reader.read(&mut b) {
+                Ok(0) => return Ok(0),
+                Err(e) => return Err(e),
+                _ => (),
+            }
+            if b != [0xfd, b'7', b'z', b'X', b'Z', 0] {
```
Comment (Pr0methean) [inline src/read/xz.rs:122]: Combine the magic bytes, flags and CRC32 into a 12-byte read, for performance reasons. You can use the `to_and_from_le!` macro to help with this. Also, please join the byte literals into a byte-slice literal: `b"\xFD7zXZ\0"`

File: src/read/xz.rs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+                written: &mut written,
+                rest: &mut self.buf,
+                total: &mut total,
+            },
+        )?;
+
+        let unpadded_size = *reader.count - block_begin;
+        self.records.push((unpadded_size, total));
+
+        if unpadded_size % 4 != 0 {
```
Comment (Pr0methean) [inline src/read/xz.rs:313]: Combine this and the following read into one, by summing the total length in a mutable variable.

**Review by Pr0methean — CHANGES_REQUESTED** [review]
Mostly looks good.

File: src/compression.rs [hunk 4]
```diff
@@ -134,6 +142,8 @@ impl CompressionMethod {
             CompressionMethod::Zstd => 93,
             #[cfg(feature = "lzma")]
             CompressionMethod::Lzma => 14,
+            #[cfg(feature = "xz")]
+            CompressionMethod::Xz => 95,
```
Comment (Pr0methean) [inline src/compression.rs:146]: Add this to `fuzz/fuzz.dict` in 16-bit little-endian form. That makes it more likely to be covered by the fuzz tests.

File: src/read/xz.rs (hunk 1 above)
Comment (yujincheng08) [reply src/read/xz.rs:71]: see line 106. it will be consumed by next call to read.

File: src/read/xz.rs [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        std::io::ErrorKind::InvalidData,
+        "Invalid multi-byte encoding",
+    ))
+}
+
+impl<R: Read> Read for XzDecoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if !self.buf.is_empty() {
+            let len = std::cmp::min(buf.len(), self.buf.len());
+            for (i, b) in self.buf.drain(..len).enumerate() {
```
Comment (Pr0methean) [inline src/read/xz.rs:106]: Combine these into one big operation using `copy_from_slice` and `VecDeque::as_slices`, for performance reasons mainly involving vectorization.

File: src/read/xz.rs [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+            }
+            let mut size = *reader.count - block_begin;
+            if size % 4 != 0 {
+                let mut b = vec![0u8; 4 - size % 4];
+                reader.read_exact(&mut b)?;
+                digest.update(&b);
+                size += b.len();
+            }
+            let mut crc = [0u8; 4];
+            reader.read_exact(&mut crc)?;
```
Comment (Pr0methean) [inline src/read/xz.rs:194]: Combine the footer reads into one as well.

**Comment by yujincheng08** [comment]
All done

**Review by Pr0methean — APPROVED** [review]

**Comment by Pr0methean** [comment]
Thanks for your contribution! I'm hoping to merge and release it within the next 24 hours.

**Comment by yujincheng08** [comment]
CI failed but it seems to come from mainline.

**Comment by yujincheng08** [comment]
@Pr0methean Hi, I found [lzma-rust](https://docs.rs/lzma-rust/0.1.6/lzma_rust/) which supports partial decompression and BCJ, and most importantly, it supports compression. Maybe you are interested in using it instead?

**Comment by yujincheng08** [comment]
I have an ugly implementation to use `lzma-rust` here: https://github.com/yujincheng08/zip2/tree/lzma-rust

But it's not panic-free so I am not opening a PR. Maybe you could build upon it.

**Comment by Pr0methean** [comment]
Opened an issue to consider `lzma-rust`. I'd still like to merge this change in its current form.

**Comment by yujincheng08** [comment]
Thanks. It is reasonable to use lzma-rs for now and later migrates to lzma-rust for both lzma and xz (and add compression support).

**Comment by Pr0methean** [comment]
Update: It may take another day or two to merge this, since I first need to generate new fuzz corpora with smaller inputs so the fuzz tests can run faster, and the scripts I'm using to do so may still need debugging.

**Comment by yujincheng08** [comment]
I dont think the repo settings are correct here.

According to the github documentation:

> The merge queue provides the same benefits as the Require branches to be up to date before merging branch protection, but does not require a pull request author to update their pull request branch and wait for status checks to finish before trying to merge.

If merge queue is set, we dont need to enable `Require branches to be up to date before merging`. But both are enabled. Thus merge cannot be done once someone push to the target branch.

Also the timeout is not correctly set while a time-consuming fuzz test is checked, and the checks nearly exceeds the time limit of Github Action (6h). If merge queue is set, some status check can be moved to merge queue while PR check can only contain neccessary ones (like fomart and compilation without unit test).

I think we could disable `Require branches to be up to date before merging`, disable unit test (at least diable fuzz test) for PR, set merge queue timeout to a longer time.

**Comment by Pr0methean** [comment]
Actually `Require branches to be up to date before merging` *isn't* enabled for this repo -- I've only been rebasing this PR when there turned out to have been a bug in master.

**Comment by Pr0methean** [comment]
As for the redundancy of the checks, that's necessary to detect the scenario where two PRs in flight at the same time are compatible with master, but not with each other.

**Pull request merged**